dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetWithVectorIndexInputStream.cpp

```diff
@@ -0,0 +1,261 @@
+            assert(vec_column->size() == file_rows);
+
+            Block block;
+            if (!rest_col_defs->empty())
+            {
+                block = readOtherColumns();
+                assert(block.rows() == vec_column->size());
+            }
+
+            auto index = header.getPositionByName(vec_cd.name);
+            block.insert(index, ColumnWithTypeAndName(std::move(vec_column), vec_cd.type, vec_cd.name));
+
+            // Fill res_filter
+            if (selected_rows == file_rows)
+            {
+                res_filter = nullptr;
+            }
+            else
+            {
+                filter.clear();
+                filter.resize_fill(file_rows, 0);
+                for (const auto & [rowid, _] : file_selected_rows)
+                    filter[rowid - read_rows] = 1;
+                res_filter = &filter;
+            }
+
+            // All rows in this ColumnFileTiny have been read.
+            block.setStartOffset(read_rows);
+            to_next_file();
+            return block;
+        }
+        auto block = (*cur_column_file_reader)->readNextBlock();
+        if (block)
+        {
+            block.setStartOffset(read_rows);
+            read_rows += block.rows();
+            res_filter = nullptr;
+            return block;
+        }
+        else
+        {
+            (*cur_column_file_reader).reset();
+            ++cur_column_file_reader;
+        }
+    }
+    return {};
+}
+
+void ColumnFileSetWithVectorIndexInputStream::load()
+{
+    if (loaded)
+        return;
+
+    tiny_readers.reserve(column_files.size());
+    UInt32 precedes_rows = 0;
+    for (const auto & column_file : column_files)
+    {
+        if (auto * tiny_file = column_file->tryToTinyFile();
+            tiny_file && tiny_file->hasIndex(ann_query_info->index_id()))
+        {
+            auto tiny_reader = std::make_shared<ColumnFileTinyVectorIndexReader>(
+                *tiny_file,
+                data_provider,
+                ann_query_info,
+                valid_rows.createSubView(precedes_rows, tiny_file->getRows()),
+                vec_cd,
+                vec_index_cache);
+            auto sr = tiny_reader->load();
+            for (auto & row : sr)
+                row.key += precedes_rows;
+            selected_rows.insert(selected_rows.end(), sr.begin(), sr.end());
+            tiny_readers.push_back(tiny_reader);
+            // avoid virutal function call
+            precedes_rows += tiny_file->getRows();
+        }
+        else
+        {
+            tiny_readers.push_back(nullptr);
+            precedes_rows += column_file->getRows();
+        }
+    }
+    // Sort by distance
+    std::sort(selected_rows.begin(), selected_rows.end(), [](const auto & lhs, const auto & rhs) {
+        return lhs.distance < rhs.distance;
+    });
+    // Only keep the top_k rows.
+    if (selected_rows.size() > ann_query_info->top_k())
+        selected_rows.resize(ann_query_info->top_k());
```

Comment:
Possibly complexity can be reduced from O(NlogN) to O(N) by using https://en.cppreference.com/w/cpp/algorithm/nth_element